### PR TITLE
Feat: expose server on local network with new --host flag

### PR DESCRIPTION
**Pull request by bholmesdev**
## Changes

- Introduce new `--host` flag to mirror [Vite's config option](https://vitejs.dev/config/#server-host)
  - When `--host` is absent, the network URL is _not_ exposed on your network
  - When `--host` is present, the network URL is exposed at a default URL
  - When `--host [custom host]` is present, the network URL is exposed at [custom host]
  - (deprecated) When `--hostname [custom host]` is present, the network URL is also exposed at [custom host]
- Deprecate `--hostname` while maintaining functionality

## Testing

- Added unit tests for current and legacy flags
- Added unit tests for `preview` command alongside `dev`

## Docs

Updated configuration and CLI references on docs.
**[PR here!](https://github.com/withastro/docs/pull/207)**

### CODE REVIEW

File: packages/astro/src/@types/astro.ts
Comment (FredKSchott) [inline packages/astro/src/@types/astro.ts:297]: needs update

File: packages/astro/src/@types/astro.ts
Comment (FredKSchott) [inline packages/astro/src/@types/astro.ts:314]: how should we handle this in the docgen script? 

I think my suggestion would be to ignoring it completely, and add a note in the `host` `@description` body that it used to be called `hostname`.

File: packages/astro/src/cli/index.ts
Comment (FredKSchott) [inline packages/astro/src/cli/index.ts:41]: nit: this feels a little verbose for just a summary. What about just `'--host [address]'` or `'--host [ip]'`? imo advanced details can be left to the documentation.

File: packages/astro/src/core/dev/index.ts
Comment (FredKSchott) [inline packages/astro/src/core/dev/index.ts:50]: I thought that `--host 1.2.3.4` would work. How does `devStart` know to show the correct `http://1.2.3.4` URL in the console log if we don't share the host string? Is `os.networkInterfaces()` that smart?

File: packages/astro/src/core/dev/util.ts
Comment (FredKSchott) [inline packages/astro/src/core/dev/util.ts:28]: NICE! I really like that this will show `localhost` instead of `127.0.0.1`

File: packages/astro/src/core/dev/util.ts
Comment (FredKSchott) [inline packages/astro/src/core/dev/util.ts:37]: does this handle if `host` is `"localhost"` or `"127.0.0.1"`?

**Review by FredKSchott — COMMENTED** [review]
Looking good! Left a few comments

File: packages/astro/src/core/dev/index.ts
Comment (bholmesdev) [reply packages/astro/src/core/dev/index.ts:50]: Yep, `os.networkInterfaces` really does cover us! This is taken directly from Vite's CLI.
For instance, passing my device IP will only expose the server on my network (not localhost or `0.0.0.0`):

Screen Shot 2022-03-11 at 9 00 53 AM

File: packages/astro/src/core/dev/util.ts
Comment (bholmesdev) [reply packages/astro/src/core/dev/util.ts:37]: It actually exposes on the network whenever `host` has a value. This mirrors Vite's `--host` flag. However, I agree that's kind of confusing! Happy to avoid exposing for these 2 values.

Comment (bholmesdev) [reply packages/astro/src/core/dev/util.ts:39]: Updated! [See here](https://github.com/withastro/astro/pull/2760/commits/f206ec5a400444970c6006b836077dd88e61c177)

File: packages/astro/src/cli/index.ts
Comment (bholmesdev) [reply packages/astro/src/cli/index.ts:41]: Fair! Just wanted to clarify that it's optional, so users know `--host` could be a boolean flag. Maybe `--host [optional IP]` is a good compromise?

File: packages/astro/src/@types/astro.ts
Comment (bholmesdev) [reply packages/astro/src/@types/astro.ts]: Made that refactor. Waiting for the docgen updates to merge, but will verify my `> blockquote` looks right once those are in 👍

**Review by natemoo-re — COMMENTED** [review]
Looks good other than a question around passing `--host 127.0.0.1` explicitly

File: packages/astro/src/core/dev/index.ts
Comment (FredKSchott) [reply packages/astro/src/core/dev/index.ts:48]: cool! TIL

File: examples/minimal/astro.config.mjs
Comment (bholmesdev) [reply examples/minimal/astro.config.mjs:10]: Aw crap, didn't mean to merge that! Will remove

File: packages/astro/src/core/messages.ts
Comment (natemoo-re) [inline packages/astro/src/core/messages.ts:10]: Super small nit: we always do `import type` at the top of the file.

No need to fix, just FYI.

**Review by natemoo-re — APPROVED** [review]
LGTM once tests are passing!

File: packages/astro/test/cli.test.js
Comment (natemoo-re) [inline packages/astro/test/cli.test.js:82]: This is failing because the `parseCliDevStart` util is pretty naive.

Comment (natemoo-re) [reply packages/astro/test/cli.test.js:82]: Fixed in [cce64e8](https://github.com/withastro/astro/pull/2760/commits/cce64e8102c485cbd6bb58d762c4b0a86857b359)

**Comment by natemoo-re** [comment]
https://media.giphy.com/media/ta83CqOoRwfwQ/giphy.gif

**Pull request merged**
